File: datasets/Part 2 - Regression/Section 6 - Polynomial Regression/polinomial_regression.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Mar  5 12:45:44 2019

@author: juangabriel and Kevin Meza
"""

####################################################################################################
#
# En una regresion poinomial se tiene una variable, pero esta esta elevada a deferentes potencias.
# y = b + b1x1 + b2x1^1 + b3x1^2 + b4x1^3 + ... + bnx1^n
# Este tipo de regresion es util cuando al graficar los datos, una curva los podria describir de mejor
# manera que una recta, como en el caso de una exponencial.
#
####################################################################################################


# REGRESIÓN POLINOMIAL

# Cómo importar las librerías
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

################################################
###          IMPORTAR EL DATA SET            ###
################################################

dataset = pd.read_csv('Position_Salaries.csv')
# En este caso no timaremos en cuenta la primer columna, pues provee la
# misma informacion que la segunda columna.

# X = dataset.iloc[:, 1].values
# Para los algoritmos de Machine Learning, se necesita una MATRIZ de caracteristicas.
# La sintaxis de arriba nos arroja un vector, mientras que al filtrar la informacion
# poniendo "1:2" en vez de "1", obtenemos como resultado una matriz.
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, 2].values

#################################################################################
### Dividir el data set en conjunto de entrenamiento y conjunto de testing    ###
#################################################################################

# En este caso no vamos a dvidir nuestro conjunto de datos, ya que contamos con
# un escaso numero de observaciones. No hay la suficiente informacion para
# entrenar un modelo. Ademas en este cas buscamos hacer una interpolacion
# para predecir el sueldo de un empleado entre estos niveles.
# Sin embargo este paso es escencial.

"""
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)
"""

################################################
#            Escalado de variables             #
################################################

# No se hara. El modelo de RLP busca entender las relaciones no lineales.
# Si se escalan los datos, esta relacion NO LINEAL podria perderse.
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)"""

##################################################
#      REFUTACION LA Ho (REGRESION LINEAL)       #
##################################################

# Generamos una regresion lineal para ver lo que pasa cuando se intenta
# ajustar datos que no son lineales. veremos despues como regresa la cosa
# con una regresion lineal polinomica.

# Ajustar la regresión lineal con todo el dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, y)

############################################################
##  Ajustar la regresión polinómica con todo el dataset    #
############################################################

from sklearn.preprocessing import PolynomialFeatures
# Seleccionamos el grado hasta el que queremos nuestra regresion pilinomica
# Por defecto es de grado 2.
poly_reg = PolynomialFeatures(degree=6)
# Transformamos nuestra matriz de caracteristicas, para generar tantas
# nuevas columnas como grados del polinomio que hayamos elegido.
# Estas nuevas columnas tendran los valores de la columna original pero
# elevados a tantas potencias, como grados tenga el polinomio.

# Transformamos nuestra matriz. La funcion añade una colunmna de 1's hasta
# el principio que servira para la obtencion del termino independiente
# en la ecuacion polinomial.
X_poly = poly_reg.fit_transform(X)
# Ahora hacemos el ajuste polinomial y creamos nustros objeto que hace
# nuestra regresion lineal. Es la misma funcion que para la multiple.
lin_reg_2 = LinearRegression()
# Ajustamos nuestra regresion a los datos.
lin_reg_2.fit(X_poly, y)

# NOTA: Hay que ir subiendo de grado el polinomio para ir generando cada
# vez un mejor ajuste hasta quedarnos con un modelo deseable.
# Cuidado con poner grados de mas.

################################################
#                PREDICCION                    #
################################################

# Finalmente averiguaremos el sueldo que se espera de un empleado de nivel
# Predicción de nuestros modelos
# PARA EL MODELO DE RLM
lin_reg.predict([[6.5]])  # Creamos un nd adarray
# PARA EL MODELO DE RP
lin_reg_2.predict(poly_reg.fit_transform([[6.5]]))


################################################
#        VISUALIZACION DE RESULTADOS           #
################################################

# Visualización de los resultados del MODELO LINEAL
plt.scatter(X, y, color="red")
plt.plot(X, lin_reg.predict(X), color="blue")
plt.title("Modelo de Regresión Lineal")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo (en $)")
plt.show()

# Pdemos ver que existe un mal ajuste de los datos.

# Visualización de los resultados del MODELO POLINÓMICO

# Para evitar que la funcion continua que predice el salario ,se vea
# como trocitos de recta, se creea una secuencia de valores que esten
# en ptos. intermedios entre el 1y el 10, que vayan de 0.1 en 0.1.
X_grid = np.arange(min(X), max(X), 0.1)
# Convertimos el vector a una matriz de 1 columna.
X_grid = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, y, color="red")
# Con esta parte hacemos una transfoemcion de la matriz de la X_grid
# que creeamos previamente y hacemos la prediccion de los valores.
plt.plot(X_grid, lin_reg_2.predict(
    poly_reg.fit_transform(X_grid)), color="blue")
plt.title("Modelo de Regresión Polinómica")
plt.xlabel("Posición del empleado")
plt.ylabel("Sueldo (en $)")
plt.show()

File: datasets/Part 2 - Regression/Section 6 - Polynomial Regression/polinomial_regression.R
# Regresión Polinómica

####################################################################################################
#
# En una regresion poinomial se tiene una variable, pero esta esta elevada a deferentes potencias.
# y = b + b1x1 + b2x1^1 + b3x1^2 + b4x1^3 + ... + bnx1^n
# Este tipo de regresion es util cuando al graficar los datos, una curva los podria describir de mejor 
# manera que una recta, como en el caso de una exponencial.
#
####################################################################################################


################################################
###          IMPORTAR EL DATA SET            ###
################################################

setwd("~/Documentos/Udemy/machinelearning-az/datasets/Part 2 - Regression/Section 6 - Polynomial Regression")
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[, 2:3]

#################################################################################
### Dividir el data set en conjunto de entrenamiento y conjunto de testing    ###
#################################################################################

# En este caso no vamos a dvidir nuestro conjunto de datos, ya que contamos con 
# un escaso numero de observaciones. No hay la suficiente informacion para
# entrenar un modelo. Ademas en este cas buscamos hacer una interpolacion
# para predecir el sueldo de un empleado entre estos niveles.
# Sin embargo este paso es escencial.

# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Purchased, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# testing_set = subset(dataset, split == FALSE)


################################################
#            Escalado de variables             #
################################################

# training_set[,2:3] = scale(training_set[,2:3])
# testing_set[,2:3] = scale(testing_set[,2:3])

##################################################
#      REFUTACION LA Ho (REGRESION LINEAL)       #
##################################################

# Generamos una regresion lineal para ver lo que pasa cuando se intenta 
# ajustar datos que no son lineales. veremos despues como regresa la cosa
# con una regresion lineal polinomica.

# Ajustar Modelo de Regresión Lineal con el Conjunto de Datos
lin_reg = lm(formula = Salary ~ ., 
             data = dataset)
summary(lin_reg)

# Con la parte de Estimate (coef.) nos podemos dar cuenta de que el modelo no es el 
# mejor. Intercept es en -195,333, ¿lo que significaria que para entrar a la empresa
# abria que pagar?. Tambien nos dice esto que al aumentar cada nivel se te
# pagan $80,879 mas, lo cual no coincide con los datos.

############################################################
##  Ajustar la regresión polinómica con todo el dataset    #
############################################################

# Ajustar Modelo de Regresión Polinómica con el Conjunto de Datos
# Le damos tratamiento a las variables independientes, construimos los terminos
# del polinomio de forma explicita, para que haya tantas columnas de la variable
# independiente como grados del polinomio.
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
dataset$Level5 = dataset$Level^5
dataset$Level6 = dataset$Level^6
poly_reg = lm(formula = Salary ~ .,
              data = dataset)
summary(poly_reg)

# NOTA: Hay que ir subiendo de grado el polinomio para ir generando cada
# vez un mejor ajuste hasta quedarnos con un modelo deseable.
# Cuidado con poner grados de mas.

################################################
#                PREDICCION                    # 
################################################

# Predicción de nuevos resultados con Regresión Lineal
# Generamos un dataframe para que lo acepte la funcion, cuya unica fila es "Level"
# (mismo nombre de la columna del dataset con el que se creo el modelo).
# Si en level colucamos un vector, podemos hacer varias predicciones a la vez.
y_pred = predict(lin_reg, newdata = data.frame(Level = c(6.5,5)))

# Predicción de nuevos resultados con Regresión Polinómica
# Hacemos un dataframe con las columnas correspondiantes a los coeficientes de
# la regresion lineal polinomica.
y_pred_poly = predict(poly_reg, newdata = data.frame(Level = 6.5,
                                                     Level2 = 6.5^2,
                                                     Level3 = 6.5^3,
                                                     Level4 = 6.5^4))

################################################
#        VISUALIZACION DE RESULTADOS           #
################################################

# Visualización del modelo lineal
# install.packages("ggplot2")
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level , y = dataset$Salary),
             color = "red") +
  # Al momento de hacer la prediccion, en "newdata", con pasarle todo el dataset
  # es  suficiente, automaticamente toma la variable independiente de definimos
  # al generar nuestro modelo de regresion lineal, unicamente la columna tiene 
  # que tener el mismo nombre que el dataset usado para crear el modelo.
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            color = "blue") +
  ggtitle("Predicción lineal del sueldo en función del nivel del empleado") +
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")


# Visualización del modelo polinómico

# Para evitar que la funcion continua que predice el salario ,se vea 
# como trocitos de recta, se creea una secuencia de valores que esten
# en ptos. intermedios entre el 1y el 10, que vayan de 0.1 en 0.1.
# Utilizando esta secuencia de numeros, generamos un nuevo dataframe 
# con tantas columnas como grados del polinomio. Nuevamente visualizamos el resultado.
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x = dataset$Level , y = dataset$Salary),
             color = "red") +
  geom_line(aes(x = x_grid, y = predict(poly_reg, 
                                        newdata = data.frame(Level = x_grid,
                                                             Level2 = x_grid^2,
                                                             Level3 = x_grid^3,
                                                             Level4 = x_grid^4,
                                                             Level5 = x_grid^5,
                                                             Level6 = x_grid^6))),
            color = "blue") +
  ggtitle("Predicción polinómica del sueldo en función del nivel del empleado") +
  xlab("Nivel del empleado") +
  ylab("Sueldo (en $)")
